### DISCUSSION TITLE
Keyboard hotkey support

### DISCUSSION THREAD

**Discussion by rafaelaroca** [BIPES/BIPES#94] [Announcements]
We have added screenshot and keyboard hotkey support on the last commit. Working hotkeys:

Ctrl+Shift+L => Export PNG image from the blocks (screenshot)
Ctrl+Shift+R => Run the program
Ctrl+Shift+E => Switch to the Console and Run the program
Ctrl+Shift+S => Stop running program

For developers: adding new hotkeys should be easy. For example:

shortcut.add("Ctrl+Shift+R",function() {
  alert("BIPES hotkey: Run program!");
  Tool.runPython();
});

The implementation was based on:
http://www.openjs.com/scripts/events/keyboard_shortcuts/

and
https://github.com/google/blockly/blob/096d1c46c5066cfa7e59db3b41405b7e854b95d0/tests/playgrounds/screenshot.js

**Comment by andrewg78** [comment]
Hi, 
If a given block has embedded image, it is not included on the screenshot.
![screenshot-test](https://user-images.githubusercontent.com/2042344/147861685-5ec08bba-694b-4bba-9d88-08f79778c566.png)

**Comment by rafaelaroca** [comment]
Hi @andrewg78. The same happens to me. I saw on the Google Blockly discussion group that other users of Blockly are also complaining about this behavior. In that way, we could create an option for "compact blocks", without images, that will be exported correctly or try to solve this issue, when possible.

**Comment by andrewg78** [comment]
Agree. Let's make an issue from this to have it in the backlog.

**Comment by runningopenloop** [comment]
Hi, so this should allow the creation of things like Ctrl+D to duplicate?
   > Hmm, I found in blockly_compressed.js:
   Blockly.duplicate=function(a){var b=Blockly.clipboardXml_,c=Blockly.clipboardSource_;Blockly.copy(a);a.workspace.paste(Blockly.clipboardXml_);Blockly.clipboardXml_=b;Blockly.clipboardSource_=c};Blockly.onContextMenu_=function(a){Blockly.utils.isTargetInput(a)||a.preventDefault()};

Ctrl+L to pop-open the Logic panel?
   > I need to do something with <category name="%{BKY_CATLOGIC}" colour="%{BKY_LOGIC_HUE}">?

Ctrl+Alt+N to create a new number?  Without needing to navigate the menu every time?
  > I need to do something to instantiate (or add to some xml?) <block type="math_number">?  How do I go about this?  I was searching add blocky block stuff, but I get a lot of noise wrt adding custom blocks.
  > So, I'm not clear what the function is which to add a new block... and I'm not experienced at all in javascript.

Is there a guide which explains the blockly code.  Again, a lot of noise wrt to how to use Blockly vs. understand the code itself...
Hmm, perhaps this is the general area: https://developers.google.com/blockly/guides/create-custom-blocks/variables#javascript,

Here's me flailing.. (actually I realize now, this new numeric variable isn't a deposited on screen as I had hoped, but rather add a new named variable... - it did popup the dialog to provide the name, but didn't actually add it.. so not the right function...

Ctrl+Shift+D and Ctrl+Alt+V seemed to do something, not what I originally intended - nor did it achieve the thing now that I realize what it does, claims to do.

	shortcut.add("Ctrl+Shift+D",function() {
		alert("BIPES hotkey: Duplicate!");
        Blockly.Variables.createVariableButtonHandler(UI ['workspace'], null, 'math_number');
	});

	shortcut.add("Ctrl+Alt+V",function() {
		alert("BIPES hotkey: New variable!");
        Blockly.Variables.createVariableButtonHandler(workspace, null, 'math_number');
	});

	shortcut.add("Ctrl+Shift+L",function() {
		alert("BIPES hotkey: Show logic!");
        Blockly.Variables.createVariableButtonHandler(Tool.getTargetWorkspace(), null, 'math_number');
	});
-----------

So yeah, I'm new to blockly too... but all this clicking is going to give me carpal tunnel or some other RSI.

73,
Timothy
